### PR TITLE
NativeAOT: ICustomMarshaller

### LINKED ISSUE #160: Implement ICustomMarshaler
Related tests:

```
Interop/ICustomMarshaler/ConflictingNames/SameNameDifferentAssembly
Interop/ICustomMarshaler/Primitives/ICustomMarshaler_TargetWindows
Interop/PInvoke/CustomMarshalers/CustomMarshalersTest
```

### PR DESCRIPTION
**Pull request by kant2002**
This is https://github.com/dotnet/runtimelab/pull/1723
Closes https://github.com/dotnet/runtimelab/issues/160

### CODE REVIEW

File: src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs [hunk 1]
```diff
@@ -536,24 +538,20 @@ protected override void EmitMarshalFieldNativeToManaged()
 
     abstract class ByValStringMarshaller : ByValArrayMarshaller
     {
-        protected override void EmitElementCount(ILCodeStream codeStream, MarshalDirection direction)
```
Comment (jkotas) [inline src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:539]: This looks like a bogus change.

Comment (kant2002) [reply src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs]: indeed :(

**Comment by kant2002** [comment]
Given that NativeAOT lives in the runtime repo, can I somehow document in the tests difference between throwing exceptions for ICustomMarshaller ?
For example I can copy `src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs ` to `src/tests/nativeaot/SmokeTests/` and made changes at that location, or maybe I can somehow add some `#define` as part of this PR, so at least differences would be tested, documented and I can cleanup `issues.targets` even more.

**Comment by MichalStrehovsky** [comment]
> Given that NativeAOT lives in the runtime repo, can I somehow document in the tests difference between throwing exceptions for ICustomMarshaller ?

How do the differences look like? One possible option would be to adjust the test to accept both exception types. We can ask the interop team what they think once we can see the exact diff. Some of the exceptions thrown for invalid inputs are really just a side effect of how things are implemented in the runtime and if it will be acceptable for NativeAOT to throw something else, it can be acceptable for any runtime and it's okay to adjust the test for all flavors.

**Comment by kant2002** [comment]
@MichalStrehovsky please take a look at the differences. I put them as comments, just so they would be visible. 
Some of them become quite generic like everything now is `MarshallingException`.
Others, like `typeof(System.Exception): Method '[ICustomMarshaler_TargetWindows]System.Runtime.InteropServices.Tests.ICustomMarshalerTests.MarshalerOnPointerMethod(int32*)' requires marshalling that is not yet supported by this compiler.` can be handled by ILC if it can recognize not only `Invalid` marshaller, but `NotImplemented`. In our case we clearly know that some marshalling is not supported and never will, but right now it falls in the "not yet implemeted" bucket.

**Comment by kant2002** [comment]
Also separate question. In order to put a block on the delegates with ref parameters I should perform some voodo. 

Right now `delegate (ref int)` => ask for marshalling of stub with this signature `{[S.P.CompilerGenerated]Internal.CompilerGenerated.<Module>.ReverseDelegateStub__TestDelegateRef(native int*)}`. Information about `ref` is lost here. It is not super important for me, but would like to make that work while I'm here.

File: src/coreclr/tools/Common/TypeSystem/Interop/IL/MarshalHelpers.cs [hunk 2]
```diff
@@ -494,6 +505,8 @@ public static partial class MarshalHelpers
                             else
                                 return MarshallerKind.ByValArray;
                         }
+                    case NativeTypeKind.CustomMarshaler:
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/TypeSystem/Interop/IL/MarshalHelpers.cs:508]: These cases (below too) are now unreachable.

File: src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs [hunk 3]
```diff
@@ -1129,4 +1131,155 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
...
+            }
+
+            var getInstanceMethod = marshallerType.GetMethod(
+                "GetInstance",
+                new MethodSignature(MethodSignatureFlags.Static, 0, customMarshallerType, new[] { Context.GetWellKnownType(WellKnownType.String) }));
+            if (getInstanceMethod == null)
+            {
+                // ApplicationException throw here.
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:1178]: This looks unnecessary - there's handling for this below and in corelib.

File: src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs [hunk 4]
```diff
@@ -1129,4 +1131,155 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
...
+                ThrowHelper.ThrowTypeLoadException(marshallerType);
+            }
+
+            var customMarshallerType = Context.SystemModule.GetKnownType("System.Runtime.InteropServices", "ICustomMarshaler");
+            if (customMarshallerType == null)
+            {
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
+
+            if (!marshallerType.CanCastTo(customMarshallerType))
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:1165]: We could do this check in CoreLib and throw the correct exception - `RuntimeImports.AreTypesAssignable` is the API to use.

File: src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs [hunk 5]
```diff
@@ -370,6 +380,7 @@ public static ICustomMarshaler GetInstance(string cookie)
 
         public static void Parameter_MarshalerOnValueType_ThrowsMarshalDirectiveException()
         {
+            // typeof(System.Exception): Method '[ICustomMarshaler_TargetWindows]System.Runtime.InteropServices.Tests.ICustomMarshalerTests.MarshalerOnValueTypeMethod(int32)' requires marshalling that is not yet supported by this compiler.
```
Comment (MichalStrehovsky) [inline src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:383]: Jeremy is fixing this one and below in #63320 (we will get rid of "requires marshalling that is not yet supported" and just throw MarshalDirective).

File: src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs [hunk 6]
```diff
@@ -1129,4 +1131,155 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
...
+            }
+
+            if (marshallerType.IsGenericDefinition)
+            {
+                // TypeLoadException throw here.
+                ThrowHelper.ThrowTypeLoadException(marshallerType);
+            }
+
+            var customMarshallerType = Context.SystemModule.GetKnownType("System.Runtime.InteropServices", "ICustomMarshaler");
+            if (customMarshallerType == null)
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:1160]: GetKnownType won't return null

File: src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs [hunk 7]
```diff
@@ -1129,4 +1131,155 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
...
+            if (MarshallerLocalVariable != (ILLocalVariable)(-1))
+            {
+                return MarshallerLocalVariable;
+            }
+
+            var marshallerType = MarshalAsDescriptor.CustomMarshallerType;
+            if (marshallerType == null)
+            {
+                // TypeLoadException throw here.
+                ThrowHelper.ThrowMarshalDirectiveException();
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:1150]: ThrowHelper.ThrowTypeLoadException?

File: src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs [hunk 8]
```diff
@@ -386,6 +398,7 @@ public static unsafe void Parameter_MarshalerOnPointer_ThrowsMarshalDirectiveExc
 
         public static void Parameter_NullICustomMarshaler_ThrowsTypeLoadException()
         {
+            // typeof(System.Runtime.InteropServices.MarshalDirectiveException): Marshaling directives are invalid.
             Assert.Throws<TypeLoadException>(() => NullCustomMarshalerMethod(""));
```
Comment (MichalStrehovsky) [inline src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:402]: I think we just have throw the right exception from the compiler (I commented on the spot).

File: src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs [hunk 9]
```diff
@@ -394,6 +407,7 @@ public static void Parameter_NullICustomMarshaler_ThrowsTypeLoadException()
 
         public static void Parameter_NotICustomMarshaler_ThrowsApplicationException()
         {
+            // typeof(System.Runtime.InteropServices.MarshalDirectiveException): Marshaling directives are invalid.
```
Comment (MichalStrehovsky) [inline src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:410]: I commented on the spot that I think is responsible for this being different.

**Review by MichalStrehovsky — COMMENTED** [review]
I think we'll be able to greatly reduce (maybe even eliminate) the number of exception differences.

File: src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs (hunk 7 above)
Comment (kant2002) [reply src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:1150]: There no type here. Best overload is `void ThrowTypeLoadException(string @namespace, string name, string moduleName)` but I have to deduce namespace and module name. Module name I can pass to `MarshalAsDescriptor` from `EcmaSignatureParser` but isn't it overkill?

Comment (MichalStrehovsky) [reply src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:1150]: If you change MarshalAsDescriptor to have a string instead of TypeDesc and do the conversion here, you can ask the conversion helper to throw exception instead of returning null. It will throw the right exception

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/InteropHelpers.cs [hunk 10]
```diff
@@ -591,6 +593,32 @@ internal static unsafe void CleanupVariant(IntPtr pDstNativeVariant)
 #endif
         }
 
+        public static unsafe object InitializeCustomMarshaller(RuntimeTypeHandle pParameterType, RuntimeTypeHandle pMarshallerType, RuntimeTypeHandle pICustomMarshallerType, string cookie, delegate*<string, object> getInstanceMethod)
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/InteropHelpers.cs:596]: `typeof(ICustomMarshaller)` or `EETypePtrOf<ICustomMarshaller>` would work just fine, no need to pass it from the generated IL.

Comment (kant2002) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/InteropHelpers.cs:596]: Excellent. I was scary to use that because you told me that Type creation is relatively expensive. So I thought this is the case too.

File: src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs (hunk 7 above)
Comment (kant2002) [reply src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs]: Look at this https://github.com/dotnet/runtime/pull/62921/commits/0fce5bae841a0f13632373eb258bd93f66461894 I take liberty there, but that's seems to be reasonable way. Notice: This case is where no `TypeDesc == null` and `typeName == null` as well.

Comment (MichalStrehovsky) [reply src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs]: Is there anything problematic with my suggestion though ("If you change MarshalAsDescriptor to have a string instead of TypeDesc and do the conversion here, you can ask the conversion helper to throw exception")? I'm not a fan of `TypeLoadException` with no message when we instead could throw "FileLoadException: assembly X could not be found", or "TypeLoadException: Type X in assembly Y could not be found".

Comment (kant2002) [reply src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs]: This is case for `null`. Null string is still `null` and we are would not have type information. Also your suggested method was my initial attempt and this is what we settle on because there no reasonable way to create TypeDesc inside Marshaller class

Comment (kant2002) [reply src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs]: That’s the case for message “Cannot load null type”

Comment (MichalStrehovsky) [reply src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs]: That field is populated as `customMarshallerType = _ecmaModule.GetTypeByCustomAttributeTypeName(managedMarshallerTypeName, false);` so this will be null for everything that didn't resolve. Not just a null string.

Maybe we can flip this to `true` and then it will never be null.

Comment (kant2002) [reply src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs]: Changes which I introduce was for this case only 
https://github.com/dotnet/runtime/blob/84926be4243143b7830e78cb7727057ed711a3cc/src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs#L387-L393

Comment (kant2002) [reply src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs]: Now I re-read and understand what you are saying.

Comment (kant2002) [reply src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs]: I did not handle just the case when type name point to separate assembly. I have to add additional test case and honestly do not want to touch that unless that's nesecssary. This seems to be fringe implementation feature. Let somebody else report issue 😄 . I would gladly help, or maybe as followup PR with tests.

**Comment by kant2002** [comment]
@MichalStrehovsky what's the plan how to move forward?

File: src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs [hunk 11]
```diff
@@ -1129,4 +1131,146 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
...
+            var customMarshallerType = Context.SystemModule.GetKnownType("System.Runtime.InteropServices", "ICustomMarshaler");
+            if (customMarshallerType == null)
+            {
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
+
+            var getInstanceMethod = marshallerType.GetMethod(
+                "GetInstance",
+                new MethodSignature(MethodSignatureFlags.Static, 0, customMarshallerType, new[] { Context.GetWellKnownType(WellKnownType.String) }));
+            if (ManagedType.IsValueType || ManagedType.IsPointer)
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:1166]:
```suggestion
            if (ManagedType.IsValueType || ManagedType.IsPointer || ManagedType.IsFunctionPointer)
```

**Review by MichalStrehovsky — COMMENTED** [review]
> @MichalStrehovsky what's the plan how to move forward?

I think it's very close, thanks for your patience! Could you also addressing the old feedback?

https://github.com/dotnet/runtime/pull/62921#discussion_r785599675 - the call to GetTypeByCustomAttributeTypeName should pass `true`. Then it will throw an exception if the type cannot be found (including handling the case when the assembly doesn't exist) and the type returned from it will never be null.

https://github.com/dotnet/runtime/pull/62921#discussion_r781703380

Once Jeremy's change gets in, we'll be able to re-test how close we're with passing the tests. So far we didn't have to do unnatural things to keep exceptional behavior the same, which is basically my bar for exception compatibility.

**Comment by kant2002** [comment]
> the call to GetTypeByCustomAttributeTypeName should pass true

I do understand that idea. I'm not convinced that this is proper fix. Maybe I miss something. Let me explain.
If I add `true` to `GetTypeByCustomAttributeTypeName` then exception would be thrown at compiler phase and I **assume** that it will crash compiler. In itself this is reasonable behavior, but this is change from CoreCLR. I would like to confirm that my assumption is correct, and ILC crash is correct inference.

Right now I handle case of incorrect type, and I even add test for that. Way how I do that seems to be fine and reasonable, so maybe I should not change what's I have right now. *If* ILC would crash ,then I should undo that test.

> Once Jeremy's change gets in

Ok. I will wait. 

So far, I only want to clarify `GetTypeByCustomAttributeTypeName` course of action and will monitor sibling PR.

**Comment by MichalStrehovsky** [comment]
Can you help me understand that concern? In your change you have:

```csharp
            if (marshallerType == null || marshallerType.IsGenericDefinition)
            {
                ThrowHelper.ThrowTypeLoadException(marshallerTypeName ?? string.Empty, module);
            }
```

That's an exception throw. Those are the same kinds of throws that will happen if you pass true to GetTypeByCustomAttributeTypeName. E.g.

https://github.com/dotnet/runtime/blob/e4bfedd847990795eb28cda4008b5ff4c58fd561/src/coreclr/tools/Common/TypeSystem/Common/Utilities/CustomAttributeTypeNameParser.cs#L255-L256

These are all caught at the same spot.

**Comment by kant2002** [comment]
@MichalStrehovsky so far still have 2 differences in same places, they are shown in tests. I test by merging Jeremy's changes from latest main.

**Comment by kant2002** [comment]
The 2b44656631eb27d71531ec366d3f1999372d42d9 irrelevant. Both variants produce same unhandled exception

**Managed exception**
```
Unhandled Exception: System.Runtime.InteropServices.MarshalDirectiveException: Marshaling directives are invalid.
   at Internal.Runtime.TypeLoaderExceptionHelper.CreateMarshalDirectiveException(ExceptionStringID) + 0x4a
   at Internal.Runtime.CompilerHelpers.ThrowHelpers.ThrowMarshalDirectiveException(ExceptionStringID) + 0x1c
   at ICustomMarshaler_TargetUnix!<BaseAddress>+0x5f6a7b
```

**Unmanaged exception**
```
> kc
 # Call Site
00 ntdll!NtRaiseException
01 KERNELBASE!RaiseFailFastException
02 ICustomMarshaler_TargetWindows!S_P_CoreLib_Interop_Kernel32__RaiseFailFastException_0
03 ICustomMarshaler_TargetWindows!S_P_CoreLib_Interop_Kernel32__RaiseFailFastException
04 ICustomMarshaler_TargetWindows!S_P_CoreLib_System_RuntimeExceptionHelpers__FailFast_1
05 ICustomMarshaler_TargetWindows!S_P_CoreLib_System_RuntimeExceptionHelpers__RuntimeFailFast
06 ICustomMarshaler_TargetWindows!S_P_CoreLib_System_Runtime_EH__UnhandledExceptionFailFastViaClasslib
07 ICustomMarshaler_TargetWindows!S_P_CoreLib_System_Runtime_EH__DispatchEx
08 ICustomMarshaler_TargetWindows!S_P_CoreLib_System_Runtime_EH__RhThrowEx
09 ICustomMarshaler_TargetWindows!RhpThrowEx
0a ICustomMarshaler_TargetWindows!S_P_CoreLib_Internal_Runtime_TypeLoaderExceptionHelper__CreateMarshalDirectiveException
0b ICustomMarshaler_TargetWindows!S_P_CoreLib_Internal_Runtime_CompilerHelpers_ThrowHelpers__ThrowMarshalDirectiveException
0c ICustomMarshaler_TargetWindows!Internal_CompilerGenerated__Module___<ReverseDelegateStub>ICustomMarshaler_TargetWindows_System_Runtime_InteropServices_Tests_ICustomMarshalerTests_TestDelegateRef
0d CustomMarshalersPrimitives!NativeParseIntDelegateRef
0e ICustomMarshaler_TargetWindows!ICustomMarshaler_TargetWindows_System_Runtime_InteropServices_Tests_ICustomMarshalerTests__CustomMarshallerWithDelegateRef
0f ICustomMarshaler_TargetWindows!ICustomMarshaler_TargetWindows_System_Runtime_InteropServices_Tests_ICustomMarshalerTests___c___DelegateParameter_MarshalerOnRefInt_ThrowsMarshalDirectiveException_b__83_0
10 ICustomMarshaler_TargetWindows!xunit_assert_Xunit_Assert__RecordException_0
11 ICustomMarshaler_TargetWindows!xunit_assert_Xunit_Assert__Throws_0<System___Canon>
12 ICustomMarshaler_TargetWindows!ICustomMarshaler_TargetWindows_System_Runtime_InteropServices_Tests_ICustomMarshalerTests__DelegateParameter_MarshalerOnRefInt_ThrowsMarshalDirectiveException
13 ICustomMarshaler_TargetWindows!ICustomMarshaler_TargetWindows_System_Runtime_InteropServices_Tests_ICustomMarshalerTests__Main
14 ICustomMarshaler_TargetWindows!ICustomMarshaler_TargetWindows__Module___MainMethodWrapper
15 ICustomMarshaler_TargetWindows!ICustomMarshaler_TargetWindows__Module___StartupCodeMain
16 ICustomMarshaler_TargetWindows!wmain
17 ICustomMarshaler_TargetWindows!invoke_main
18 ICustomMarshaler_TargetWindows!__scrt_common_main_seh
19 ICustomMarshaler_TargetWindows!__scrt_common_main
1a ICustomMarshaler_TargetWindows!wmainCRTStartup
1b KERNEL32!BaseThreadInitThunk
1c ntdll!RtlUserThreadStart
```

**Comment by kant2002** [comment]
Add test for that specific scenario https://github.com/dotnet/runtime/pull/64137

File: src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs [hunk 12]
```diff
@@ -658,15 +677,15 @@ private class InnerCustomMarshaler : ILargeInterface, ICustomMarshaler
...
-        [return: MarshalAs(UnmanagedType.CustomMarshaler, MarshalTypeRef = typeof(BoxedValueTypeCustomMarshaler))]
         public delegate string TestDelegateRef([MarshalAs(UnmanagedType.CustomMarshaler, MarshalTypeRef = typeof(BoxedValueTypeCustomMarshaler))] ref int val);
 
         [DllImport("CustomMarshalersPrimitives", EntryPoint = "NativeParseIntDelegateRef")]
-        [return: MarshalAs(UnmanagedType.CustomMarshaler, MarshalTypeRef = typeof(BoxedValueTypeCustomMarshaler))]
         public static extern string CustomMarshallerWithDelegateRef(int val, TestDelegateRef dlg);
 
         public static void DelegateParameter_MarshalerOnRefInt_ThrowsMarshalDirectiveException()
         {
+            // throw MarshalDirectiveException, but it is unhandled (?)
```
Comment (MichalStrehovsky) [inline src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:687]: We should probably fix this so that GetFunctionPointerForDelegate doesn't return "broken" function pointers that are just going to throw. Should be a separate pull request. Let's not enable this test for NativeAOT for now.

File: src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs [hunk 13]
```diff
@@ -596,6 +614,7 @@ public class ThrowingCleanUpNativeDataCustomMarshaler : ICustomMarshaler
 
         public static void Field_ParentIsStruct_ThrowsTypeLoadException()
         {
+            //  typeof(System.Runtime.InteropServices.MarshalDirectiveException): Struct 'StructWithCustomMarshalerField' requires marshalling that is not yet supported by this compiler.
```
Comment (MichalStrehovsky) [inline src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:617]: I think it should be fine to update this test to also accept MarshalDirectiveException. TypeLoadException is odd.

**Review by MichalStrehovsky — COMMENTED** [review]
Looks good otherwise.

@jkoritzinsky could you please have a look at the changes to the tests?

File: src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs (hunk 13 above)
Comment (kant2002) [reply src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs]: I change this exception type in this PR. that was fairly trivial. I can move that change to separate PR if needed.

Comment (MichalStrehovsky) [reply src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs]: Definitely a separate pull request. I somewhat doubt changing the exception thrown out of CoreCLR will be accepted, but one can always try.

Comment (kant2002) [reply src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:617]: Remove code. Somebody has to rerun pipeline

File: src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs (hunk 12 above)
Comment (kant2002) [reply src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:687]: Do this problem have tracking issue? I would like to change issue which prevent test from running to something different to indicate that https://github.com/dotnet/runtimelab/issues/160 indeed closed.

Comment (kant2002) [reply src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:687]: Can I hide that test under 
```
if (RuntimeInformation.OSPlatform == Platform.Windows && !RuntimeFeatured.IsDynamicCodeSupported)
```

that way I will capture your words in the test. Maybe just 
```
if (!RuntimeFeatured.IsDynamicCodeSupported)
```

File: src/coreclr/System.Private.CoreLib/System.Private.CoreLib.csproj [hunk 14]
```diff
@@ -280,8 +280,8 @@
     <Compile Include="$(BclSourcesRoot)\System\Runtime\InteropServices\CustomMarshalers\EnumerableViewOfDispatch.cs" />
     <Compile Include="$(BclSourcesRoot)\System\Runtime\InteropServices\CustomMarshalers\EnumeratorToEnumVariantMarshaler.cs" />
     <Compile Include="$(BclSourcesRoot)\System\Runtime\InteropServices\CustomMarshalers\EnumeratorViewOfEnumVariant.cs" />
-    <Compile Include="$(BclSourcesRoot)\System\Runtime\InteropServices\CustomMarshalers\ExpandoToDispatchExMarshaler.cs" />
-    <Compile Include="$(BclSourcesRoot)\System\Runtime\InteropServices\CustomMarshalers\TypeToTypeInfoMarshaler.cs" />
+    <Compile Include="$(CommonPath)\System\Runtime\InteropServices\CustomMarshalers\ExpandoToDispatchExMarshaler.cs" />
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/System.Private.CoreLib.csproj:283]: I do not think we want to add these to NativeAOT just to make some random tests pass. 

These are part of built-in COM interop support. We do not have built-in support in NativeAOT. It does not make sense to add a small incoherent bits of it.

File: src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs (hunk 12 above)
Comment (kant2002) [reply src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:687]: I think I rushing again. I agree with Michal, `GetFunctionPointerForDelegate` should throw if somebody pass delegate which contains broken marshaling patterns. That would automatically fix current issue because throw happens before delegate was passed to native code.

My idea to create(if needed) separate issue, and re-assign test to that issue, because right now all related tests for  ICustomMarshaler passing. This problem in my understanding related to any delegates marshaling which contains unsupported configuration.

Comment (jkotas) [reply src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:687]: It would be best to use `src\libraries\Common\tests\TestUtilities\System\PlatformDetection.cs` for this.

Comment (jkotas) [reply src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:687]: > I agree with Michal, GetFunctionPointerForDelegate should throw if somebody pass delegate which contains broken marshaling patterns.

FWIW, it is not how CoreCLR works. CoreCLR will throw only once you actually try to call the delegate.

File: src/coreclr/System.Private.CoreLib/System.Private.CoreLib.csproj (hunk 14 above)
Comment (kant2002) [reply src/coreclr/System.Private.CoreLib/System.Private.CoreLib.csproj:283]: So I should undo changes and mark test as CoreCLR specific?

Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/System.Private.CoreLib.csproj]: It would be best to mark it as built-in COM interop specific using `PlatformDetection.SupportsComInterop`.

Comment (kant2002) [reply src/coreclr/System.Private.CoreLib/System.Private.CoreLib.csproj]: Thank you.

File: src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs (hunk 12 above)
Comment (kant2002) [reply src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:687]: I cannot add `src\libraries\Common\tests\TestUtilities\System\PlatformDetection.cs` alone because it complains about `IsWindows` properties. If I add `src\libraries\Common\tests\TestUtilities\System\PlatformDetection.Windows.cs` test build complain about `GenerateDllImportAttribute`. If I add reference to `TestUtilities.csproj` then `src\tests\build.sh` complain about not restored assets for the project.

I look for usages of this project in the tests, but seems to be it is used only in libraries tests. Please advice.

Comment (jkotas) [reply src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:687]: If it is hard to use the full implementation from libraries currently, I would create a simplified local implementation of  PlatformDetection for runtime tests, with the subset of properties, and use that instead.

@jkoritzinsky Do you happen to have anything better in the plan for the runtime test consolidation?

Comment (jkoritzinsky) [reply src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:687]: I think for now, making a copy either locally or in the `CoreCLRTestLibrary` assembly is the way to go. As we move tests over, we might decide to consolidate some of these utility libraries.

File: src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs [hunk 15]
```diff
@@ -699,7 +718,11 @@ public static int Main(String[] args)
                 Parameter_CleanUpNativeDataMethodThrows_ThrowsActualException();
                 Field_ParentIsStruct_ThrowsTypeLoadException();
                 Parameter_DifferentCustomMarshalerType_MarshalsCorrectly();
-                DelegateParameter_MarshalerOnRefInt_ThrowsMarshalDirectiveException();
+                if (SupportsComInterop)
```
Comment (jkoritzinsky) [inline src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:721]: This feels like the wrong name for this condition. Maybe `SupportsExceptionInterop` would be a better name?

File: src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs (hunk 13 above)
Comment (jkoritzinsky) [reply src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs:618]: Is this comment still necessary?

**Review by jkoritzinsky — APPROVED** [review]
LGTM with a few nits

**Comment by kant2002** [comment]
I have enough endorphin supplies for today, but maybe this PR can also be merged? I seems to be resolve all issues from interested parties.

File: src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs (hunk 6 above)
Comment (jkotas) [reply src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:1168]: This does not look addressed.

**Comment by jkotas** [comment]
> I seems to be resolve all issues from interested parties.

There two small comments do not appear to be addressed:
- https://github.com/dotnet/runtime/pull/62921#discussion_r781703380
- https://github.com/dotnet/runtime/pull/62921#discussion_r796082492

**Review by jkotas — APPROVED** [review]
LGTM. Thank you!
